src/libraries/System.IO.FileSystem/tests/Base/BaseGetSetUnixFileMode.cs

```diff
@@ -0,0 +1,233 @@
+
+        // When false, the Get API returns (UnixFileMode)(-1) when the platform is not supported.
+        protected virtual bool GetModeThrowsPNSE => true;
+
+        // Determines if the derived Test class is for directories or files.
+        protected virtual bool IsDirectory => false;
+
+        // Linux doesn't support setting UnixFileMode on links.
+        private bool SetModeSupportsLink => PlatformDetection.IsBsdLike;
+
+        // On OSX, directories created under /tmp have same group as /tmp.
+        // Because that group is different from the test user's group, chmod
+        // returns EPERM when trying to setgid on directories, and for files
+        // chmod filters out the bit.
+        // We skip the tests with setgid.
+        private bool CanSetGroup => !PlatformDetection.IsBsdLike;
+
+        private string CreateTestItem(string path = null, [CallerMemberName] string memberName = null, [CallerLineNumber] int lineNumber = 0)
+        {
+            path = path ?? GetTestFilePath(null, memberName, lineNumber);
+            if (IsDirectory)
+            {
+                Directory.CreateDirectory(path);
+            }
+            else
+            {
+                File.Create(path).Dispose();
+            }
+            return path;
+        }
+
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        [Theory]
+        [MemberData(nameof(TestUnixFileModes))]
+        public void SetThenGet(UnixFileMode mode)
+        {
+            if (!CanSetGroup && (mode & UnixFileMode.SetGroup) != 0)
+            {
+                return; // Skip
+            }
+            if (GetModeNeedsReadableFile)
+            {
+                // Ensure the file remains readable.
+                mode |= UnixFileMode.UserRead;
+            }
+
+            string path = CreateTestItem();
+
+            SetMode(path, mode);
+
+            Assert.Equal(mode, GetMode(path));
+        }
+
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        [Theory]
+        [MemberData(nameof(TestUnixFileModes))]
+        public void SetThenGet_SymbolicLink(UnixFileMode mode)
+        {
+            if (!CanSetGroup && (mode & UnixFileMode.SetGroup) != 0)
+            {
+                return; // Skip
+            }
+            if (GetModeNeedsReadableFile)
+            {
+                // Ensure the file remains readable.
+                mode |= UnixFileMode.UserRead;
+            }
+
+            string path = CreateTestItem();
+
+            string linkPath = GetTestFilePath();
+            File.CreateSymbolicLink(linkPath, path);
+
+            if (ApiFollowsLink)
+            {
+                SetMode(linkPath, mode);
+
+                Assert.Equal(mode, GetMode(linkPath));
+                Assert.Equal(mode, GetMode(path));
+            }
+            else
+            {
+                if (!SetModeSupportsLink)
+                {
+                    Assert.Throws<IOException>(() => SetMode(linkPath, mode));
```

Comment:
This IOException indicates Linux doesn't support the operation. PNSE is probably a preferable exception type for that?

Reply:
I agree